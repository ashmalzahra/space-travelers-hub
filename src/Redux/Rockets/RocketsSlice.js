import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const url = 'https://api.spacexdata.com/v3/rockets';

const initialState = {
  rockets: [],
  ifSucceed: false,
  isLoading: false,
  error: null,
};

const fetchRockets = createAsyncThunk(
  'rockets/fetchRockets',
  async () => {
    try {
      const { data } = await axios.get(url);
      return data;
    } catch (error) {
      return error.message;
    }
  },
);

const RocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    reserveRocket: (state, action) => {
      const newArray = state.rockets.map((item) => {
        if (item.rocket_id === action.payload) {
          return { ...item, reserved: true };
        }
        return item;
      });
      return { ...state, rockets: newArray };
    },
    cancelReserveRocket: (state, action) => {
      const newArray = state.rockets.map((item) => {
        if (item.rocket_id === action.payload) {
          return { ...item, reserved: false };
        }
        return item;
      });
      return { ...state, rockets: newArray };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchRockets.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(fetchRockets.fulfilled, (state, action) => ({
        ...state,
        isLoading: false,
        ifSucceed: true,
        rockets: action.payload,
      }))
      .addCase(fetchRockets.rejected, (state) => ({
        ...state,
        isLoading: false,
      }));
  },
});

export const { reserveRocket, cancelReserveRocket } = RocketsSlice.actions;
export { fetchRockets };
export default RocketsSlice.reducer;
