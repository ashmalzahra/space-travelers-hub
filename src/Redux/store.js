import { configureStore } from '@reduxjs/toolkit';
import rocketsReducer from './RocketsSlice';
import missionsSlice from './MissionsSlice';
// import missionsReducer from "./MissionsSlice";

const store = configureStore({
  reducer: {
    rockets: rocketsReducer,
    missions: missionsSlice.reducer,
  },
});

export default store;
