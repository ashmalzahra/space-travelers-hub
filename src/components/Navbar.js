import { NavLink } from 'react-router-dom';
import Logo from '../Images/planet.png';

function Navbar() {
  return (
    <div className="NavBar">
      <div className="LogoContainer">
        <img className="planet" src={Logo} alt="planet-logo" />
        <h1>Traverlers Hub</h1>
      </div>
      <nav className="NavLinks">
        <NavLink to="/" end className={({ isActive }) => (isActive ? 'currentPage' : undefined)}>Rockets</NavLink>
        <NavLink to="/missions" end className={({ isActive }) => (isActive ? 'currentPage' : undefined)}>Missions</NavLink>
        <NavLink to="/profile" end className={({ isActive }) => (isActive ? 'currentPage' : undefined)}>My Profile</NavLink>
      </nav>
    </div>
  );
}

export default Navbar;
