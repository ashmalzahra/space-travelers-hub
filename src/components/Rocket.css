.container {
  display: flex;
  flex-direction: row;
  margin-top: 4%;
  margin-bottom: 4%;
}

.pic {
  max-width: 275px;
}

.title {
  margin-left: 2.5%;
}

.description {
  margin-left: 2.5%;
  font-size: 20px;
}

.btn-primary,
.btn-light {
  margin-left: 2.5%;
}

.btn-success {
  margin-right: 1.5%;
}
