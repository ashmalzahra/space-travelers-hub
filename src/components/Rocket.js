import React from 'react';

const Rocket = ({id, name, type, image}) => {

    return(
            <div>
                <img src={image} alt={name}/>
                <div>
                    <h3>{name}</h3>
                    <p>{type}</p>
                    <button id={id}>Reserve Rocket</button>
                </div>
            </div>
        )
}

export default Rocket