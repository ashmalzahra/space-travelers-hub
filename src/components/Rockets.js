import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets } from '../Redux/RocketsSlice';
import Rocket from './Rocket';

const Rockets = () => {
  const dispatch = useDispatch();
  const ifSucceed = useSelector((store) => store.rockets.ifSucceed);
  const isLoading = useSelector((store) => store.rockets.isLoading);
  const rockets = useSelector((store) => store.rockets.rockets);

  useEffect(() => {
    dispatch(fetchRockets());
  }, [ifSucceed, dispatch]);

  let content;
  if (isLoading) {
    content = (
      <div>
        <p>Loading...</p>
      </div>
    );
  } else if (ifSucceed) {
    content = rockets.map((item) => (
      <Rocket
        key={item.rocket_id}
        rocket={item}
      />
    ));
  } else {
    content = (
      <div>
        <p>Something went wrong...</p>
      </div>
    );
  }
  return (
    <div>{content}</div>
  );
};

export default Rockets;
