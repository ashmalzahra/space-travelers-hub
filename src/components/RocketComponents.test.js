import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import store from '../Redux/store';
import Rockets from './Rockets';
import Navbar from './Navbar';
import Rocket from './Rocket';
import Joined from './RocketProfile';

const obj = {
  rocket_name: 'Falcon 1',
  rocket_id: 'falcon1',
};

test('Rockets rendered correctly', () => {
  const configureStore = store;
  const component = renderer.create(
    <Provider store={configureStore}>
      <MemoryRouter>
        <Rockets />
      </MemoryRouter>
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Navbar renders correctly', () => {
  const configureStore = store;
  const component = renderer.create(
    <Provider store={configureStore}>
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Rocket reserved displayed in profile', () => {
  const configureStore = store;
  const component = renderer.create(
    <Provider store={configureStore}>
      <MemoryRouter>
        <Joined />
      </MemoryRouter>
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Rocket renders as expected', () => {
  const configureStore = store;
  const component = renderer.create(
    <Provider store={configureStore}>
      <MemoryRouter>
        <Rocket key={obj.rocket_id} rocket={obj} />
      </MemoryRouter>
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
