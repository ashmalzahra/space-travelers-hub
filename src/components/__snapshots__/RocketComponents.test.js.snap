// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<div
  className="NavBar"
>
  <div
    className="LogoContainer"
  >
    <img
      alt="planet-logo"
      className="planet"
      src="planet.png"
    />
    <h1>
      Traverlers Hub
    </h1>
  </div>
  <nav
    className="NavLinks"
  >
    <a
      aria-current="page"
      className="currentPage"
      href="/"
      onClick={[Function]}
    >
      Rockets
    </a>
    <a
      href="/missions"
      onClick={[Function]}
    >
      Missions
    </a>
    <a
      href="/profile"
      onClick={[Function]}
    >
      My Profile
    </a>
  </nav>
</div>
`;

exports[`Rocket renders as expected 1`] = `
<div
  className="container"
  id="falcon1"
>
  <img
    alt="Falcon 1"
    className="pic"
  />
  <div>
    <h3
      className="title"
    >
      Falcon 1
    </h3>
    <p
      className="description"
    />
    <button
      className="btn btn-primary"
      disabled={false}
      id="falcon1"
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;

exports[`Rocket reserved displayed in profile 1`] = `
<div
  className="reserve-container"
  id="container"
>
  <h1
    className="profile-missions-title"
  >
    My Rockets
  </h1>
  <ul
    className="profile-missions"
  >
    <p
      className="no-reserve"
    >
      No Rockets Reserved
    </p>
  </ul>
</div>
`;

exports[`Rockets rendered correctly 1`] = `
<div>
  <div>
    <p>
      Something went wrong...
    </p>
  </div>
</div>
`;
